### PR TITLE
SNOW-761229 Bug fix: encoding ocsp request data in URL to avoid invalid URL issue

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-hx**
https://github.com/snowflakedb/snowflake-sdks-drivers-issues-teamwork/issues/259
Encode Base64 OCSP Request data in URL for some corner cases that it might have characters that are not URL safe.
Unit test only as it's for for some corner cases that not easy to reproduce.

### CODE REVIEW

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c [hunk 1]
```diff
@@ -638,9 +687,10 @@ static OCSP_RESPONSE * queryResponderUsingCurl(char *url, OCSP_CERTID *certid, c
   {
     if (!ACTIVATE_SSD)
     {
+      encodeUrlData(ocsp_req_base64, ocsp_req_base64_len, &encoded_ocsp_req_base64);
       snprintf(urlbuf, sizeof(urlbuf),
                ocsp_cache_server_retry_url_pattern,
-               host, ocsp_req_base64);
+               host, encoded_ocsp_req_base64);
```
Comment (sfc-gh-ext-simba-hx) [inline deps/curl-7.88.1/lib/vtls/sf_ocsp.c:693]: Checked entire sf_ocsp.c this seems to be the only place put base64 encoded ocsp request data into URL.

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c [hunk 2]
```diff
@@ -266,6 +268,53 @@ static const unsigned int MAX_SLEEP_TIME = 16000000;
...
 
 
+/* Encode request data as a part of URL
+ * Encode all characters reserved by URL
+ * @param url_data The source data needs to be encoded_ptr
+ * @param data_size The size of the source data
+ * @param encoded_ptr The allocated buffer filled with encoded data, need to be freed by caller
+ * return the size of encoded data
+ */
+size_t encodeUrlData(const char *url_data, size_t data_size, char** encoded_ptr)
```
Comment (sfc-gh-igarish) [inline deps/curl-7.88.1/lib/vtls/sf_ocsp.c:278]: what will happen URL is already encoded?

Comment (sfc-gh-ext-simba-hx) [reply deps/curl-7.88.1/lib/vtls/sf_ocsp.c:278]: That won't happen as here we only encode the base64 request data that will be appended as a part of the URL, not entire URL.
encodeUrlData(ocsp_req_base64, ocsp_req_base64_len, &encoded_ocsp_req_base64);

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c [hunk 3]
```diff
@@ -266,6 +268,53 @@ static const unsigned int MAX_SLEEP_TIME = 16000000;
...
+{
+  // allocate buffer 3 times larger than source data in case every single, add 1 for '\0'
+  // character needs to be encoded as %xx
+  size_t buf_len = data_size * 3 + 1;
+  char* encode_buf = (char*)malloc(buf_len);
```
Comment (sfc-gh-igarish) [inline deps/curl-7.88.1/lib/vtls/sf_ocsp.c:283]: Can we check for null?

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c [hunk 4]
```diff
@@ -758,6 +809,7 @@ static OCSP_RESPONSE * queryResponderUsingCurl(char *url, OCSP_CERTID *certid, c
 
 end:
   if (ocsp_req_base64) curl_free(ocsp_req_base64);
+  if (encoded_ocsp_req_base64) curl_free(encoded_ocsp_req_base64);
```
Comment (sfc-gh-igarish) [inline deps/curl-7.88.1/lib/vtls/sf_ocsp.c:812]: You allocated using malloc and then free using curl_free. is that ok?

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c (hunk 3 above)
Comment (sfc-gh-igarish) [inline deps/curl-7.88.1/lib/vtls/sf_ocsp.c:283]: Make sure malloc is right API here and during free it calls curl_free.

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c [hunk 5]
```diff
@@ -266,6 +268,53 @@ static const unsigned int MAX_SLEEP_TIME = 16000000;
...
+      snprintf(cur_ptr, buf_len, "%%%.2X", car);
+      cur_ptr += 3;
+      enc_len += 3;
+      buf_len -= 3;
+    }
+  }
+
+  *cur_ptr = '\0';
+  *encoded_ptr = encode_buf;
+  return enc_len;
```
Comment (sfc-gh-igarish) [inline deps/curl-7.88.1/lib/vtls/sf_ocsp.c:315]: Are we using in the caller function?

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c (hunk 3 above)
Comment (sfc-gh-ext-simba-hx) [reply deps/curl-7.88.1/lib/vtls/sf_ocsp.c]: check added

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c (hunk 4 above)
Comment (sfc-gh-ext-simba-hx) [reply deps/curl-7.88.1/lib/vtls/sf_ocsp.c:832]: This is following the existing curl implementation here:
https://github.com/snowflakedb/libsnowflakeclient/blob/master/deps/curl-7.88.1/lib/base64.c#L206
The buffer ocsp_req_base64 is allocated with malloc and freed using curl_free as well.

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c (hunk 3 above)
Comment (sfc-gh-ext-simba-hx) [reply deps/curl-7.88.1/lib/vtls/sf_ocsp.c]: Confirmed, following existing curl implementation.

File: deps/curl-7.88.1/lib/vtls/sf_ocsp.c (hunk 5 above)
Comment (sfc-gh-ext-simba-hx) [reply deps/curl-7.88.1/lib/vtls/sf_ocsp.c]: Not really but won't hurt to have data length returned. Make it optional output argument.

**Review by sfc-gh-igarish — APPROVED** [review]
LGTM
